src/EFCore.PG/Metadata/PostgresEnum.cs

```diff
@@ -9,102 +9,163 @@
+        [NotNull]
         public static PostgresEnum GetOrAddPostgresEnum(
             [NotNull] IMutableAnnotatable annotatable,
             [CanBeNull] string schema,
             [NotNull] string name,
             [NotNull] string[] labels)
         {
+            Check.NotNull(annotatable, nameof(annotatable));
+            Check.NullButNotEmpty(schema, nameof(schema));
+            Check.NotEmpty(name, nameof(name));
+            Check.NotNull(labels, nameof(labels));
+
             if (FindPostgresEnum(annotatable, schema, name) is PostgresEnum enumType)
                 return enumType;
 
-            enumType = new PostgresEnum(annotatable, BuildAnnotationName(schema, name));
-            enumType.SetData(labels);
-            return enumType;
-        }
+            var annotationName = BuildAnnotationName(schema, name);
 
-        public static PostgresEnum GetOrAddPostgresEnum(
-            [NotNull] IMutableAnnotatable annotatable,
-            [NotNull] string name,
-            [NotNull] string[] labels)
-            => GetOrAddPostgresEnum(annotatable, null, name, labels);
+            return new PostgresEnum(annotatable, annotationName) { Labels = labels };
+        }
 
+        /// <summary>
+        /// Finds a <see cref="PostgresEnum"/> in the <see cref="IAnnotatable"/>, or returns null if not found.
+        /// </summary>
+        /// <param name="annotatable">The annotatable to search for the enum.</param>
+        /// <param name="schema">The enum schema or null to use the model's default schema.</param>
+        /// <param name="name">The enum name.</param>
+        /// <returns>
+        /// The <see cref="PostgresEnum"/> from the <see cref="IAnnotatable"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException"><paramref name="schema"/></exception>
+        /// <exception cref="ArgumentNullException"><paramref name="annotatable"/></exception>
+        /// <exception cref="ArgumentNullException"><paramref name="name"/></exception>
+        [CanBeNull]
         public static PostgresEnum FindPostgresEnum(
             [NotNull] IAnnotatable annotatable,
             [CanBeNull] string schema,
             [NotNull] string name)
         {
             Check.NotNull(annotatable, nameof(annotatable));
+            Check.NullButNotEmpty(schema, nameof(schema));
             Check.NotEmpty(name, nameof(name));
 
             var annotationName = BuildAnnotationName(schema, name);
 
             return annotatable[annotationName] == null ? null : new PostgresEnum(annotatable, annotationName);
         }
 
+        [NotNull]
         static string BuildAnnotationName(string schema, string name)
-            => NpgsqlAnnotationNames.EnumPrefix + (schema == null ? name : schema + '.' + name);
-
+            => schema != null
+                ? $"{NpgsqlAnnotationNames.EnumPrefix}{schema}.{name}"
+                : $"{NpgsqlAnnotationNames.EnumPrefix}{name}";
```

Comment:
Good idea. Will have to punt for 2.2.1 though.

Reply:
Hey @AndriySvyryd :) Didn't see your comment before submitting my review below...

* Any specific reason for to have a single annotation for all enums?
* The current scheme is already in place, so unless I'm mistaken a change here would break existing migration code...
* Is annotation lookup/traversal considered a perf-sensitive area? I was under the impression it's only used at initialization/migration generation time...

Reply:
> I haven't checked how this is used, but potentially the annotation could be looked up every time a value of this type is sent to the database.

I'm not sure how this could be happening (unless I'm forgetting something)... At least for this specific annotation - define an enum type - this is purely an annotation on the model itself, which should only be picked up at the migrations SQL generation level for creating the appropriate PostgreSQL enum type... The other part of the enum support is at the TypeMappingSource, but that happens as a context option and isn't related to the model or to any annotations.

But we should definitely be sure of the perf situation here...

Regardless, as @austindrenski said, it's a bit late to do this for 2.2.0 (we're planning to release in a few hours :)). But still definitely important...

Reply:
If it's just migrations (like `Sequence`) then you are correct that it doesn't matter much, but it might be worthwhile to check where this and other annotations that have constructed names are used.

Reply:
OK, thanks for the tip - we'll definitely take a look at this after the release to make sure.

Reply:
Opened #736 to track further discussion after `2.2.0`.